Search busking

pray92 opened the pull request:
Implementation of the entity of busking and getting list of busking ready to start and starting. Offset range is Korea. If client tries to search the list of busking off from Korea. It returns exception.

Simple search logics are implemented. Detail implementation will be implemented after this branch.

pray92 committed:
FEAT: Busking Entity, CUD Repository and test case

pray92 committed:
FEAT: Get DAO for getting activated busking list

pray92 committed:
FEAT: Calculator for coordinate

pray92 committed:
FEAT: DAO for buskings finding

pray92 committed:
FEAT: Controller for busking view in map

- If its range is off-korea, throw exception

f-lab-nolan commented on src/main/java/kr/texturized/muus/application/service/BuskingViewService.java:
```diff
@@ -0,0 +1,35 @@
+     * @return list of active busking information for map
+     */
+    public List<BuskingMapVo> getActiveBuskingsInMap(
+        double latitude,
```
항상 습관적으로 final을 붙여사용하는게 좋아요.

pray92 replied:
방심했습니다. 반영했습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/application/service/calculator/CoordinateCalculator.java:
```diff
@@ -0,0 +1,69 @@
+
+        double width = meterToLatitude(widthMeter);
+        double height = meterToLongitude(heightMeter);
+        if (MIN_LATITUDE > latitude - width * 0.5
```
여기는 주석으로 간단한 설명을 넣는게 좋을것 같아요.
어떤 기준으로 판별하는지에 대해서요

pray92 replied:
반영했습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/dao/BuskingFindDao.java:
```diff
@@ -0,0 +1,43 @@
+@RequiredArgsConstructor
+public class BuskingFindDao {
+
+    private final CoordinateCalculator coordinateCalculator;
```
하위 패키지에서 상위 패키지 참조는 절대 하지 않아요.
추후에] 개발 시에 순환참조가 발생할 수 있어요.
CoordinateCalculator 클래스가 공용으로 쓸 수 있다면
별도의 패키지로 빠지는게 좋을 것 같아요.

pray92 replied:
이부분은 application 계층에서 쓸것이라 예상하고 초기에 패키징했었는데, 개발하는 과정에서 놓쳤던 부분이었습니다. application.calculator에서 common.coordinate로 리패키징하였습니다.

f-lab-nolan commented on src/main/java/kr/texturized/muus/domain/entity/Busking.java:
```diff
@@ -0,0 +1,91 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Getter
+@ToString(of = {"host", "title", "createTime"})
+public class Busking {
```
보통 기술 면접 시
jpa N+1 문제, jpa 양방향 순환참조 문제  등에 대해
해결 방안을 물어볼 수 있어요.
찾아서 정리 및 이해가 필요해요.

또한 entity에 @OneToMany 대신에 
각 엔티티를 각각 조회하는거에 대해서도 고민해 보세여.
ex)
// 주문 정보
Order order = repository.findById();
// 주문에 해당하는 상품 정보
List<Product> list = repository.findByOrderId(orer.id());

pray92 replied:
그러한 접근 방법이 있었겠군요. 이부분이 언급하신 내용이랑 관련있는 것도 인지했습니다. JPA N+1 문제, 양방향 순환 참조를 추후 학습 후 다음 branch에서 개선하도록 하겠습니다(다음 branch도 busking 관련 feature입니다).

f-lab-nolan commented on src/test/java/kr/texturized/muus/application/service/calculator/CoordinateCalculatorTest.java:
```diff
@@ -0,0 +1,51 @@
+    @Test
+    void calculateCoordinate() {
+        double latitude = calculator.meterToLatitude(1_000_000.0);
+        double longitude = calculator.meterToLongitude(1_000_000.0);
```
결과에 대한 검증이 없는데.. 
어떤 의도가 있는걸가요?

pray92 replied:
추후 개발하는 과정에서 각 위도와 경도의 미터 단위가 궁금할 때마다 사용하기 위해 두었습니다. 약간 고민이 들지만....아무래도 불필요한 테스트 로직이 발생하니 빼두는 것이 좋을까요?

f-lab-nolan replied:
어떤 의도였는지 언제까지 유지할건지등의 내용을 주석으로
남겨놓아야 할것 같아요. 
그래야 나중에 지워도 되는 코드인지 어떤 코드인지 알수 있을테니까요

pray92 replied:
존중해주셔서 감사합니다. 반영했습니다.

f-lab-nolan commented on src/test/java/kr/texturized/muus/application/service/calculator/CoordinateCalculatorTest.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    void expectInvalid(double latitude, double longitude, double widthMeter, double heightMeter) {
+        try {
```
try~catch가 없어야 하지 않을가요?
calculator.validateRange 에서 exception이 발생하면 
아무 일이 생기지 않는데요.

invalidRangeThenExpectError 테스트 메소드는 
Assertions.assertThrows(OutOfCoordinateRangeException.class, () -> {  expectInvalid(0.0, 0.0, 1, 1);
  });
이런식으로 해야하지 않을가요?

pray92 replied:
throw 관련 테스트 메서드 사용법이 미숙했습니다. 반영했습니다.

pray92 committed:
CHORE: add final keyword

pray92 committed:
CHORE: add comments for coordinate validation

pray92 committed:
CHORE: Throw test using JUnit module

pray92 committed:
CHORE: Move Coordinate modules application -> common.coordinate

pray92 committed:
CHORE: Add comment of test for coordinate calculation

- It seems meaningless, but it's easy to calculate latitude and longitude
- Let it leave for development
